### DISCUSSION TITLE
Auth mechanism in tutorial unsafe?

### DISCUSSION THREAD

**Discussion by gbuerk** [noam-honig/radweb#79] [General]
Hey all,

I was thinking about the authentication flow in the tutorial and it occurred to me that having the `AuthController.ts` file in the shared file seems insecure.  Doesn't shared code get sent to the client?  And yes, it turns out it does.  I can see, in my browser, the default JWT secret: 
```
return (await import('jsonwebtoken')).sign(user, process.env['JWT_SECRET'] || "my secret")
```
That's maybe fine as long as the environment variable is utilized, but seems not great practice to me.  Even if we removed the "my secret" default, it's probably not great practice to leak the process used to authenticate.

As an alternative, we can use a traditional express route that is server side only.

Thoughts?

**Comment by yoni-rapoport** [comment]
Not sure if having standard authentication code sent to the client is unsafe or a bad practice. (The default JWT secret is obviously unsafe and only there for the tutorial to work without environment variables set up.)

A traditional express route is an option, but there are ways to get both the type-safety of a @BackendMethod call and an [implementation that isn't shared with the client](https://remult.dev/docs/using-server-only-packages.html#solution-2-abstract-the-call).

I think it's better to keep the tutorial authentication in its current simple form, and add a some text pointing out the fact that this code is sent to clients and explaining the alternatives in case the reader isn't comfortable with that.

**Reply by gbuerk** [reply]
That's a great point about keeping type-safety and abstracting the server-only code.  I tried an implementation using the server injection technique and it works beautifully.  That's definitely better than a traditional express route.  I'm loving this project!  

So, I'm on board with maybe just a little text in the tutorial pointing out that the code is sent to the client.  When writing a real app, I was tempted to put my auth logic (which delegates to another auth server) right in the AuthController which, in my case, would have been insecure.
